compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -1185,6 +1186,32 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
       patch(Position(toUntyped(vdef).pos.start), "@volatile ")
   }
 
+  /** Adds inline to final vals with idempotent rhs
+   *
+   *  duplicating scalac behavior: for final vals that have rhs as constant, we do not create a field
+   *  and instead return the value. This seemingly minor optimization has huge effect on initialization
+   *  order and the values that can be observed during superconstructor call
+   *
+   *  see remark about idempotency in PostTyper#normalizeTree
+   */
+  private def patchFinalVals(vdef: ValDef)(implicit ctx: Context): Unit = {
+    def skipBlocks(t: Tree): Tree = t match {
+      case Block(_, t1) => skipBlocks(t1)
+      case _ => t
+    }
+    def isFinalInlinableVal(sym: Symbol): Boolean = {
+      sym.is(Final, butNot = Mutable) &&
+      isIdempotentExpr(vdef.rhs) /* &&
+      ctx.scala2Mode (stay compatible with Scala2 for now) */
+    }
+    val sym = vdef.symbol
+    skipBlocks(vdef.rhs) match {
```

Comment:
I think the right body is:

       val sym = vdef.symbol
       sym.info match {
          case info: ConstantType if isFinalInlinableVal(sym) => sym.flags |= Inline
          case _ =>
       }